### ISSUE TITLE
Delegate Missing Values and Categorical Handling in `GrouperTransformer` and `GrouperPredictor`

### ISSUE THREAD

**Issue by sergiocalde94** [koaning/scikit-lego#656]
**Overview**

I'd like to request/ask for enhanced functionality for the `GrouperTransformer` and `GrouperPredictor` classes to delegate handling of missing values and categorical features directly to the specified transformer/estimator. This feature would align `scikit-lego` with the native handling capabilities of popular machine learning libraries like CatBoost and LightGBM, and categorical encoding libraries like `category_encoders`.

**Use Case**

The main advantage of this feature is that it leverages the native support for missing values and categorical features present in many modern (and widely used) transformers and estimators. Instead of preprocessing the data separately, this allows users to directly pass through missing values and categorical features. Here's a concrete example illustrating the proposed enhancement:

```python
from sklego.preprocessing import ColumnSelector
from sklego.meta import GroupedTransformer, GroupedPredictor
from category_encoders.james_stein import JamesSteinEncoder
from catboost import CatBoostClassifier
from sklearn.pipeline import make_pipeline


# Feature selection
column_selector = ColumnSelector(NUM_FEATURES + CAT_FEATURES)

# Categorical encoder
cat_encoder = JamesSteinEncoder(cols=CAT_FEATURES)

# Grouped Transformer by country
cat_encoder_by_country = GroupedTransformer(
    transformer=cat_encoder, groups="country"
)

# CatBoost Classifier setup
cb = CatBoostClassifier(
    num_trees=300,
    max_depth=3,
    colsample_bylevel=0.75,
    min_child_samples=50,
    reg_lambda=150,
    verbose=0
)

# Grouped Predictor by country
cb_by_country = GroupedPredictor(
    estimator=cb, groups="country"
)

# Final pipeline
pipeline_cb = make_pipeline(
    column_selector, cat_encoder_by_country, cb_by_country
)

# Train the pipeline
pipeline_cb.fit(X_train, y_train)
```

This is not working now, no matter if the argument `check_X` is `True` or `False`.

**Comment by FBruzzesi** [comment]
Hey there! I tried to replicate it with some data, but for estimators with native NaN support I was not able to get an error.

Example:

```py
import numpy as np
import pandas as pd

from sklearn.ensemble import HistGradientBoostingRegressor
from sklego.meta import GroupedPredictor

target = "y"

frame = pd.DataFrame({
    "country": list("abcd") * 25,
    "x": np.random.randn(100),
}).assign(
    y = lambda t: t["x"]*2,
    z = lambda t: np.where(np.abs(t["x"])>1, np.nan, t["x"])
)

X, y = frame.drop(columns=[target]), frame[target]
#  X.isna().sum() # sanity check to see that `z` has missing values

model = GroupedPredictor(
    estimator=HistGradientBoostingRegressor(),
    groups="country",
    check_X=False,
)

model.fit(X, y).predict(X)  # Runs without issues
```

**Comment by FBruzzesi** [comment]
I will definitely need to take a closer look at the issue, but for now I can offer a few workarounds:

- Move to [`HierarchicalClassifier`](https://koaning.github.io/scikit-lego/api/meta/#sklego.meta.hierarchical_predictor.HierarchicalClassifier)/[`HierarchicalRegressor`](https://koaning.github.io/scikit-lego/api/meta/#sklego.meta.hierarchical_predictor.HierarchicalRegressor) (depending on the task)
- Instead of wrapping the meta model into a pipeline, I would opt for the other way around, wrap the pipeline into the meta estimator.
- The previous point has the benefit of moving everything inside a single meta estimator.

Here is a working example with the mentioned adjustments:

```py
import numpy as np
import pandas as pd

from category_encoders.james_stein import JamesSteinEncoder

from sklearn import clone
from sklearn.ensemble import HistGradientBoostingRegressor
from sklearn.pipeline import make_pipeline

from sklego.meta import HierarchicalRegressor

target = "y"

frame = pd.DataFrame({
    "country": list("abcd") * 25,
    "x": np.random.randn(100),
}).assign(
    y = lambda t: t["x"]*2,
    z = lambda t: np.where(np.abs(t["x"])>1, np.nan, t["x"]),
    zz = lambda t: np.where(np.abs(t["x"])>1, "this", "fail")
)

X, y = frame.drop(columns=[target]), frame[target]

cat_encoding = JamesSteinEncoder(cols=["zz"])

estimator = make_pipeline(
    cat_encoding, HistGradientBoostingRegressor(),
)

model = HierarchicalRegressor(
    estimator=estimator,
    groups="country",
    check_X=False,    
)

preds = model.fit(X, y).predict(X)
```

Sanity check:
```py
preds_loop = np.zeros(X.shape[0])

for name, grp in frame.groupby("country"):
    idx = grp.index

    X_grp, y_grp = grp.drop(columns=["country", target]), grp[target]
    preds_loop[idx] = clone(estimator).fit(X_grp, y_grp).predict(X_grp)
    
np.allclose(preds, preds_loop) # True
```

**Comment by sergiocalde94** [comment]
Thanks for your answer :) This approach is working but not sure yet why this is working with the hierarchical classifier but not with the grouper classifier. Now I am only splitting by country and I don't have any hierarchy but I think it is working as it should.

Offtopic on this issue but still a concern: Is there any way of tuning the parameters for each group? Also having the possibility of using different types of models would be awesome, don't know if this is something that you have in mind.

I would love to help but not sure if this is something simple to check/add.

**Comment by FBruzzesi** [comment]
>  this is working with the hierarchical classifier but not with the grouper classifier

We got a few issue with grouped predictor, and to keep backward compatibility while adding features, we decided to introduce a new meta estimator. There should be nothing that `GroupedPredictor` does and `HierarchicalPredictor`s do not.

> Is there any way of tuning the parameters for each group? Also having the possibility of using different types of models would be awesome, don't know if this is something that you have in mind.

I am afraid that the answer is negative to both with the current implementation. Also being a quite niche use case I am not sure we want to maintain something that allows that (I can see a lot of edge cases).

I would suggest a couple of options:

1. Wrap a model/estimator on something which is tunable at fit time to dynamically select the best model
2. Bring your own meta estimator, a draft of a naive implementation would be something along the following:

```py
class SwitchEstimator(BaseEstimator, RegressorMixin):

    def __init__(self, estimator_mapping, group, fallback_estimator = None):
        self.estimator_mapping = estimator_mapping
        self.group = group
        self.fallback_estimator = fallback_estimator

    def fit(self, X, y):

        ... # some preprocessing steps
        frame = pd.DataFrame(X).assign(y=y)

        self.estimators_ = {
            self.estimator_mapping.get(name, self.fallback_estimator).fit(grp.drop(columns="y"), grp["y"])
            for name, grp in frame.groupby(self.group)
        }
        return self

    def predict(self, X):
        ...
```

**Comment by FBruzzesi** [comment]
I took a closer look at the issue: the problem lies in the fact that `GroupedPredictor` does not maintain column names for the values (only for the groups). Therefore category encoders doesn't find the column (`zz` in our case) and it enters the following raise block of [`BaseEncoder`](https://github.com/scikit-learn-contrib/category_encoders/blob/master/category_encoders/utils.py#L313). 

Again, `HierarchicalPredictor`s fix that problem by keeping the original column names. I would argue that's the way to go.
